lib/common/abstract-descriptor.h

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_COMMON_ABSTRACT_DESCRIPTOR_H_
+#define FORTRAN_COMMON_ABSTRACT_DESCRIPTOR_H_
+
+#include "constexpr-bitset.h"
+
+namespace Fortran::common {
+
+// DescriptorInterface both serves to describe and validate the interface
+// that descriptor implementations must comply with in order to be used to
+// instantiate Fortran transformational intrinsic abstract implementation.
+// To validate the compliance of a descriptor implementation DESC with the
+// interface, one must compile in a test file:
+//
+//  template class DescriptorInterface<DESC>;
+//  template class Transformational<DescriptorInterface<DESC>>;
+//
+// The first line ensures DESC has all properties required by the descriptor
+// interface. The second line ensures the Abstract Fortran Runtime
+// implementation is only using descriptor properties that are part of the
+// descriptor interface. These are static tests only and it does not validate
+// the descriptor implementation logic.
+// Note that inheritance and abstract functions were not selected to make the
+// interface to avoid runtime overhead with virtual functions.
+
+template<typename DESC> class DescriptorInterface : private DESC {
+public:
+  // A descriptor implementation must define the following type
+
+  // basic integer types
+  using SubscriptValue = typename DESC::SubscriptValue;
+  using SizeValue = typename DESC::SizeValue;  // may or may not be unsigned
+  using Rank = typename DESC::Rank;
```

Comment:
Rank is a value in [0..15] and `int` would work just fine.  No need to abstract it.

Reply:
I abstracted it because of `CFI_rank_t` that aliases `unsigned char` in include/flang/ISO_Fortran_binding.h. However, I agree it is a bit overkill and that the interface  should not have to care about the types used by the implementation to store the information.

I have updated the PR to remove `Rank` and use `int` directly as suggested.